flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/BufferResponseDecoderDelegate.java

```diff
@@ -0,0 +1,117 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.netty;
+
+import org.apache.flink.runtime.io.network.NetworkClientHandler;
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandlerContext;
+
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.BufferResponse;
+
+/**
+ * The parser for {@link org.apache.flink.runtime.io.network.netty.NettyMessage.BufferResponse}.
+ */
+public class BufferResponseDecoderDelegate implements NettyMessageDecoderDelegate {
+
+	/** The network client handler of current channel. */
+	private final NetworkClientHandler networkClientHandler;
+
+	/** The Flink Buffer allocator. */
+	private final NetworkBufferAllocator allocator;
+
+	/** The cumulation buffer of message header. */
+	private ByteBuf messageHeaderBuffer;
+
+	/**
+	 * The current BufferResponse message that are process the buffer part.
+	 * If it is null, we are still processing the message header part, otherwise
+	 * we are processing the buffer part.
+	 */
+	private BufferResponse currentResponse;
+
+	/** How much bytes have been received or discarded for the buffer part. */
+	private int decodedBytesOfBuffer;
+
+	public BufferResponseDecoderDelegate(NetworkClientHandler networkClientHandler) {
+		this.networkClientHandler = networkClientHandler;
+		this.allocator = new NetworkBufferAllocator(networkClientHandler);
+	}
+
+	@Override
+	public void onChannelActive(ChannelHandlerContext ctx) {
+		messageHeaderBuffer = ctx.alloc().directBuffer(BufferResponse.MESSAGE_HEADER_LENGTH);
+	}
+
+	@Override
+	public void startParsingMessage(int msgId, int messageLength) {
+		currentResponse = null;
+		decodedBytesOfBuffer = 0;
+
+		messageHeaderBuffer.clear();
+	}
+
+	@Override
+	public ParseResult onChannelRead(ByteBuf data) throws Exception {
+		if (currentResponse == null) {
+			ByteBuf toDecode = ByteBufUtils.cumulate(messageHeaderBuffer, data, BufferResponse.MESSAGE_HEADER_LENGTH);
+
+			if (toDecode != null) {
+				currentResponse = BufferResponse.readFrom(toDecode, allocator);
+
+				if (currentResponse.bufferSize == 0) {
+					return ParseResult.finishedWith(currentResponse);
+				}
+			}
+		}
+
+		if (currentResponse != null) {
+			boolean isDiscarding = allocator.isPlaceHolderBuffer(currentResponse.getBuffer());
+			int remainingBufferSize = currentResponse.bufferSize - decodedBytesOfBuffer;
+			int actualBytesToDecode = Math.min(data.readableBytes(), remainingBufferSize);
+
+			if (isDiscarding) {
+				data.readerIndex(data.readerIndex() + actualBytesToDecode);
+			} else {
+				currentResponse.getBuffer().asByteBuf().writeBytes(data, actualBytesToDecode);
+			}
+
+			decodedBytesOfBuffer += actualBytesToDecode;
+
+			if (decodedBytesOfBuffer == currentResponse.bufferSize) {
+				if (isDiscarding) {
+					networkClientHandler.cancelRequestFor(currentResponse.receiverId);
+					return ParseResult.finishedWith(null);
```

Comment:
Based on the [comment](https://github.com/apache/flink/pull/7368/files#r381770789), we can also reuse `ParseResult.finishedWith(currentResponse)` directly to avoid bringing `null` message.

Reply:
Removed the `null` branch